[GeoMechanicsApplication] Towards element permeability matrices

avdg81 opened the pull request:
**📝 Description**
Pre-calculate relative permeability values as preparation for calculating element permeability matrices. When needed, modify these values by accounting for the permeability update factors (i.e. to account for volume changes in geometrically linear elements).

**🆕 Changelog**
Other changes include:
- Members `CalculateRetentionResponse` no longer compute the relative permeability. It is now pre-calculated.
- Extracted members that calculate the relative permeability values for all integration points of an element.
- Moved the calculation of permeability update factors to the transport equation utilities.
- Extracted a utility function that calculates fluid pressures at all integration points of an element.
- Inlined the wrapper member functions that calculate and add the permeability matrix, to avoid needless levels of indirection.
- Removed member `PermeabilityUpdateFactor` from two `ElementVariables` data structures.
- Removed the unused `ProcessInfo` data member from the retention laws.
- Removed several comments that had no additional value.
- Miscellaneous cleanup:
  * Reduced the scope of some variables.
  * Don't use `noalias` if no performance benefit is to be expected.
  * Prefer to use `this->` to refer to base members rather than a `using` statement.
  * Removed some unnecessary `KRATOS_TRY` and `KRATOS_CATCH` statements.
  * Removed several empty statements.
  * Unnamed some unused function parameters.

avdg81 committed:
Inlined the members that calculate and add the permeability matrices

avdg81 committed:
Removed ref to `ProcessInfo` object from `RetentionLaw::Parameters`

The reference has been removed, since it was not used. Consequently,
some other function signatures could be simplified, too.

avdg81 committed:
Extracted a member that computes the relative permeability values

avdg81 committed:
Extracted a utility function that calculates all fluid pressures

As a result, the calculation of the relative permeability values could
be simplified, since it will receive them rather than calculate them.

avdg81 committed:
Precalculate the relative permeability values

So far, this is done for the small strain (updated Lagrangian) U-Pw
elements (with and without FIC extension), the (updated Lagrangian) U-Pw
diff order elements, and the steady-state / transient Pw elements.

avdg81 committed:
Removed a few unnecessary `using` statements

As a result, some member function calls needed an explicit `this->`.

avdg81 committed:
Removed a few references to the permeability update factor

It has no meaning in the context of transient/steady state Pw elements.

avdg81 committed:
Started to combine the relative permeability and its update factor

avdg81 committed:
Extracted a member for calculating all permeability update factors

Also marked a related member `[[nodiscard]]`.

avdg81 committed:
Accounted for the permeability update factors

The relative permeability factors now implicitly account for the
permeability factors.

avdg81 committed:
Removed member `PermeabilityUpdateFactor` from `ElementVariables`

avdg81 committed:
Moved the calculation of the permeability update factor(s)

The member functions that calculate the permeability update factors have
been moved from classes `UPwSmallStrainElement` and
`SmallStrainUPwDiffOrderElement` to `GeoTransportEquationUtilities`.
This reduces the number of duplicated lines.

avdg81 committed:
Removed two variants of `CalculatePermeabilityMatrix`

These variants had an explicit function parameter for passing the
permeability update factor. Since we account for that by modifying the
value of the relative permeability, these variants are no longer needed.

avdg81 committed:
Removed member `PermeabilityUpdateFactor` from `ElementVariables`

The permeability update factor is now accounted for by modifying the
relative permeability values.

avdg81 committed:
Completed moving the calculation of the relative permeability values

It is no longer carried out by `CalculateRetentionResponse`.

avdg81 committed:
Clean up modified code

- Removed several redundant comments.
- Added a few blank lines where needed.
- Added a few missing `[[nodiscard]]` markers.
- Removed some empty statements.
- Removed a few redundant `KRATOS_TRY` and `KRATOS_CATCH` statements.
- Unnamed a few unused function parameters.
- Corrected a typo.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_updated_lagrangian_FIC_element.cpp:
```diff
@@ -108,6 +109,7 @@ void UPwUpdatedLagrangianFICElement<TDim, TNumNodes>::CalculateAll(MatrixType& r
         this->CalculateShapeFunctionsSecondOrderGradients(FICVariables, Variables);
 
         this->CalculateRetentionResponse(Variables, RetentionParameters, GPoint);
+        Variables.RelativePermeability = relative_permeability_values[GPoint];
```
This picks up the values computed before the i.p. loop. Surprisingly ( for me at least ) I don't see the computation disappearing within the loop. Was there no computation present?

avdg81 replied:
Yes, there was, but it was hidden inside of `CalculateRetentionResponse` (therefore, I've put the new line right below it). If you look there, you will see that the calculation of the relative permeability has been removed.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_updated_lagrangian_element.cpp:
```diff
@@ -99,6 +100,7 @@ void UPwUpdatedLagrangianElement<TDim, TNumNodes>::CalculateAll(MatrixType& rLef
             Variables.BodyAcceleration, Variables.NContainer, Variables.VolumeAcceleration, GPoint);
 
         this->CalculateRetentionResponse(Variables, RetentionParameters, GPoint);
+        Variables.RelativePermeability = relative_permeability_values[GPoint];
```
Same question as for the FIC element. Now the RelativePermeability is computed and put the variables bucket. Before it seems not to be there. Is this a fixed bug?

avdg81 replied:
No, it used to be calculated here as well (but again inside of `CalculateRetentionResponse`). What might be a bug in the FIC element are the missing permeability update factors. I will add an issue on our product backlog for that.

WPK4FEM commented on applications/GeoMechanicsApplication/custom_elements/steady_state_Pw_piping_element.cpp:
```diff
@@ -233,7 +232,7 @@ void SteadyStatePwPipingElement<TDim, TNumNodes>::CalculateAll(MatrixType& rLeft
             Variables.LocalPermeabilityMatrix, Variables.JointWidth, Prop[TRANSVERSAL_PERMEABILITY]);
 
-        CalculateRetentionResponse(Variables, RetentionParameters, GPoint);
+        this->CalculateRetentionResponse(Variables, RetentionParameters, GPoint);
```
Because the using statement dropped from the .hpp file.

avdg81 replied:
Yes, that is correct.

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp:
```diff
@@ -978,6 +970,13 @@ void UPwSmallStrainElement<TDim, TNumNodes>::CalculateAll(MatrixType&        rLe
                                          strain_vectors, mStressVector, constitutive_matrices);
     const auto biot_coefficients = GeoTransportEquationUtilities::CalculateBiotCoefficients(
         constitutive_matrices, this->GetProperties());
+    auto relative_permeability_values = this->CalculateRelativePermeabilityValues(
+        GeoTransportEquationUtilities::CalculateFluidPressures(Variables.NContainer, Variables.PressureVector));
+    const auto permeability_update_factors = GeoTransportEquationUtilities::CalculatePermeabilityUpdateFactors(
+        strain_vectors, this->GetProperties());
+    std::transform(relative_permeability_values.cbegin(), relative_permeability_values.cend(),
+                   permeability_update_factors.cbegin(), relative_permeability_values.begin(),
+                   std::multiplies{});
```
Are the permeability update factors also needed separately here? Otherwise I'd propose to extract a utility function with these lines if it makes sense

avdg81 replied:
No, the permeability update factors are not needed separately here. I can imagine us defining an overload of `CalculateRelativePermeabilityValues` that accepts the update factors as a separate parameter. Then we could also do the transformation inside that new member. I would, however, wait a little longer, just to see whether that is what we need.

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_small_strain_element.cpp:
```diff
@@ -1028,54 +1027,31 @@ std::vector<array_1d<double, TDim>> UPwSmallStrainElement<TDim, TNumNodes>::Calc
 
         FluidFluxes.push_back(PORE_PRESSURE_SIGN_FACTOR * Variables.DynamicViscosityInverse *
-                              Variables.RelativePermeability * Variables.PermeabilityUpdateFactor *
+                              Variables.RelativePermeability *
```
This means the relative permeability now always implicitly includes the update factor. Was the relative permeability used like this in every location?

avdg81 replied:
I have checked all usages of both `RelativePermeability` and `PermeabilityUpdateFactor`, and as far as I could see this was the case.

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/U_Pw_updated_lagrangian_FIC_element.cpp:
```diff
@@ -85,6 +84,8 @@ void UPwUpdatedLagrangianFICElement<TDim, TNumNodes>::CalculateAll(MatrixType& r
                                          strain_vectors, mStressVector, constitutive_matrices);
     const auto biot_coefficients =
         GeoTransportEquationUtilities::CalculateBiotCoefficients(constitutive_matrices, Prop);
+    const auto relative_permeability_values = this->CalculateRelativePermeabilityValues(
+        GeoTransportEquationUtilities::CalculateFluidPressures(Variables.NContainer, Variables.PressureVector));
```
I think I heard you mention this, but why don't we need the permeability update factors for updated lagrangian elements?

avdg81 replied:
If I remember rightly, this has to do with the fact that Updated Lagrangian elements implicitly account for deformations in the adopted strain measures. As such, the effect of deformation on the relative permeability is already captured. This is not the case for geometrically linear elements, hence the correction using permeability update factors which are computed from the strain vectors.

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_element.cpp:
```diff
@@ -1032,13 +1030,14 @@ void SmallStrainUPwDiffOrderElement::CalculateOnIntegrationPoints(const Variable
+            const auto relative_permeability =
+                mRetentionLawVector[GPoint]->CalculateRelativePermeability(RetentionParameters) *
+                GeoTransportEquationUtilities::CalculatePermeabilityUpdateFactor(
+                    Variables.StrainVector, GetProperties());
```
Is there a specific reason we cannot extract a list here?

avdg81 replied:
No, there is not. I have changed the loop such that we precalculate the relative permeability values.

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/small_strain_U_Pw_diff_order_element.cpp:
```diff
@@ -1047,8 +1046,7 @@ void SmallStrainUPwDiffOrderElement::CalculateOnIntegrationPoints(const Variable
             AuxFluidFlux        = PORE_PRESSURE_SIGN_FACTOR * Variables.DynamicViscosityInverse *
-                           RelativePermeability * Variables.PermeabilityUpdateFactor *
-                           prod(Variables.IntrinsicPermeability, GradPressureTerm);
+                           relative_permeability * prod(Variables.IntrinsicPermeability, GradPressureTerm);
```
Might be GitHub, but it seems this is not aligned correctly, could you try clang-format?

avdg81 replied:
When I reformat using clang-format, nothing changes. I do agree with you that the result looks a bit weird.

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/transient_Pw_element.cpp:
```diff
@@ -303,30 +302,22 @@ void TransientPwElement<TDim, TNumNodes>::InitializeSolutionStep(const ProcessIn
 
 //----------------------------------------------------------------------------------------------------
 template <unsigned int TDim, unsigned int TNumNodes>
-void TransientPwElement<TDim, TNumNodes>::InitializeNonLinearIteration(const ProcessInfo& rCurrentProcessInfo)
```
Not for this PR, but does it make sense that the Pw version does nothing here (i.e. ignores the UPw functionality)

avdg81 replied:
Good question. Perhaps we can discuss this later with @WPK4FEM?

rfaasse commented on applications/GeoMechanicsApplication/custom_elements/transient_Pw_element.cpp:
```diff
@@ -537,12 +529,16 @@ template <unsigned int TDim, unsigned int TNumNodes>
+    const auto permeability_matrix = GeoTransportEquationUtilities::CalculatePermeabilityMatrix<TDim, TNumNodes>(
+        rVariables.GradNpT, rVariables.DynamicViscosityInverse, rVariables.PermeabilityMatrix,
+        rVariables.RelativePermeability, rVariables.IntegrationCoefficient);
+    rLeftHandSideMatrix += permeability_matrix;
```
I see this is the same as it was, but do you know why in this case we use += and in the UPw element we do something like an assemble?

avdg81 replied:
I agree that this is equivalent to what you see in the U-Pw element. But I'm not sure why it was decided to write it in this way here.

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/transport_equation_utilities.hpp:
```diff
@@ -153,6 +150,41 @@ class GeoTransportEquationUtilities
+            result.emplace_back(CalculateFluidPressure(row(rNContainer, i), rPressureVector));
+        }
+        return result;
+    }
```
In my opinion it would be clearer to move this one closer to `CalculateFluidPressure`

avdg81 replied:
I think it was, but after resolving merge conflicts, I believe they got separated. Thanks for bringing this to my attention.

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/transport_equation_utilities.hpp:
```diff
@@ -153,6 +150,41 @@ class GeoTransportEquationUtilities
+        }
+
+        return 1.0;
+    }
```
Is this one ever used separately? If not, it could be private (maybe now the same holds for `CalculateFluidPressure`?)

avdg81 replied:
There used to be one such usage (see above), but after processing your suggestion (which was to pre-calculate the relative permeability values), it vanished. So I have made it `private` now. Good catch!

rfaasse replied:
Nice! And the `CalculateFluidPressure`? Is that still used separately, or only via `CalculateFluidPressures`?

avdg81 replied:
It appears to be still used in other places, too. Perhaps this is something that we can look into later?

rfaasse commented on applications/GeoMechanicsApplication/custom_utilities/transport_equation_utilities.hpp:
```diff
@@ -153,6 +150,41 @@ class GeoTransportEquationUtilities
+            return CalculatePermeabilityUpdateFactor(rStrainVector, rProperties);
+        });
+        return result;
+    }
```
Could you add some unit tests for these new functions?

avdg81 replied:
Done.

rfaasse reviewed (commented):
Nice clean-up, we're untangling the variables step by step! I have some suggestions of things that we might be able to clean up even more after this PR and there are still some unit tests missing for the new utilities, but I think we're almost there!

WPK4FEM reviewed (dismissed):
Good to get rid of the unnecessarily passed project info.
Further I can follow the setup of the permeability matrix.

avdg81 committed:
Processed review comments by Richard

- Precalculate the relative permeability values when calculating values
  at integration points.
- Made member `CalculatePermeabilityUpdateFactor` `private` since it was
  only used internally.
- Moved the definition of `CalculateFluidPressures` directly adjacent to
  the one of `CalculateFluidPressure`.
- Added a unit test for utility function `CalculateFluidPressures`.

avdg81 committed:
Processed more review comments by Richard

- Added three unit tests for utility function
  `CalculatePermeabilityUpdateFactors`.
- Added a constant value for the tolerance when asserting numerical
  values.

rfaasse commented on applications/GeoMechanicsApplication/tests/cpp_tests/test_transport_equation_utilities.cpp:
```diff
@@ -134,4 +141,67 @@ KRATOS_TEST_CASE_IN_SUITE(CalculateBiotCoefficients_GivesInfResults_ForZeroBulkM
+                              expected_factors, tolerance)
+}
+
+KRATOS_TEST_CASE_IN_SUITE(PermeabilityUpdateFactorEqualsOneWhenChangeInverseFactorIsNonPositive, KratosGeoMechanicsFastSuite)
+{
+    const auto unused_strain_vectors = std::vector<Vector>(3, Vector{});
+    auto       properties            = Properties{};
+    properties[PERMEABILITY_CHANGE_INVERSE_FACTOR] = -1.0;
+
+    const auto expected_factors = std::vector<double>(unused_strain_vectors.size(), 1.0);
+    KRATOS_EXPECT_VECTOR_NEAR(GeoTransportEquationUtilities::CalculatePermeabilityUpdateFactors(
+                                  unused_strain_vectors, properties),
+                              expected_factors, tolerance)
+
+    properties[PERMEABILITY_CHANGE_INVERSE_FACTOR] = 0.0;
+
+    KRATOS_EXPECT_VECTOR_NEAR(GeoTransportEquationUtilities::CalculatePermeabilityUpdateFactors(
+                                  unused_strain_vectors, properties),
+                              expected_factors, tolerance)
+}
```
Here this might be a similar question for a next issue: where will we check the PERMEABILITY_CHANGE_INVERSE_FACTOR and is this expected behavior for the function

avdg81 replied:
I honestly don't know... Perhaps this is something that we may need to discuss with the rest of the team?

rfaasse replied:
I created an issue for checking some other user defined material parameters, so we include it there 👍

rfaasse reviewed (dismissed):
Thanks for incorporating the review comments and adding the unit tests. This seems to me like it's ready to go!

avdg81 committed:
Merge remote-tracking branch 'origin/master' into geo/towards-element-permeability-matrices
